feat(common): introduce experimental `httpResource`

alxhub opened the pull request:
`httpResource` is a new frontend to the `HttpClient` infrastructure. It declares a dependency on an HTTP endpoint. The request to be made can be reactive, updating in response to signals for the URL, method, or otherwise. The response is returned as an instance of `HttpResource`, a `WritableResource` with some additional signals which represent parts of the HTTP response metadata (status, headers, etc).

alxhub committed:
feat(common): introduce experimental `httpResource`

`httpResource` is a new frontend to the `HttpClient` infrastructure. It
declares a dependency on an HTTP endpoint. The request to be made can be
reactive, updating in response to signals for the URL, method, or otherwise.
The response is returned as an instance of `HttpResource`, a
`WritableResource` with some additional signals which represent parts of the
HTTP response metadata (status, headers, etc).

Saurav-Sharma22 reviewed (approved):
approved

JeanMeche commented:
About the Adev failure, this is again due to #54858 and the workaround is the same as #58367 (commenting ADEV tests).

JeanMeche commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,364 @@
+        abortSignal.addEventListener('abort', onAbort);
+
+        // Start off stream as undefined.
+        const stream = signal<{value: T} | {error: unknown}>({value: undefined as T});
```
Nit: 
```suggestion
        const stream = signal<ResourceStreamItem<T>>({value: undefined as T});
```

shannonvc commented on packages/common/http/test/resource_spec.ts:
```diff
@@ -0,0 +1,201 @@
+    expect(res.value()).toEqual('[1,2,3]');
+  });
+
+  it('should support text responses', async () => {
```
This test is a duplicate of the one above.

alxhub committed:
ci: disable adev tests

The changes in #58288 are responsible for breaking the tests. (see #54858).
We'll re-enable them on the next release.

josephperrott reviewed (approved):
LGTM

jelbourn commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,460 @@
+   *
+   * @experimental
+   */
+  arrayBuffer: {
```
There's a lot of duplication between each sub-method here (`arrayBuffer`, `block`, etc.). Would it make sense to try to extract and parameterize this type?

alxhub replied:
I did that originally, but it doesn't work with the support we have in adev for this style of API (function with sub-functions). I'm following Paul's guidance here on how to structure this type to get the documentation to generate correctly.

We can expand adev's support for such APIs (and we definitely should).

jelbourn commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,460 @@
+  const injector = options?.injector ?? inject(Injector);
+
+  const toHttpRequest = () => {
+    let unwrappedRequest = typeof request === 'function' ? request() : request;
+    if (unwrappedRequest === undefined) {
+      return undefined;
+    } else if (typeof unwrappedRequest === 'string') {
+      unwrappedRequest = {url: unwrappedRequest};
+    }
```
optional: could refactor an `unwrapRequest` / `normalizeRequest` function out of here

```ts
function normalizeRequest(rawRequest: string | ...) {
  const unwrapped = typeof rawRequest === 'function' ? rawRequest() : rawRequest;
  return typeof unwrapped === 'string' ? {url: unwrappedRequest} : unwrapped;
}

// ...
const request = normalizeRequest(rawRequest);
if (request === undefined) {
  return undefined;
}
```

jelbourn commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,460 @@
+      unwrappedRequest.method ?? 'GET',
+      unwrappedRequest.url,
+      unwrappedRequest.body ?? null,
+      {
+        headers:
+          unwrappedRequest.headers instanceof HttpHeaders
+            ? unwrappedRequest.headers
+            : new HttpHeaders(
+                unwrappedRequest.headers as
+                  | Record<string, string | number | Array<string | number>>
+                  | undefined,
+              ),
+        params:
+          unwrappedRequest.params instanceof HttpParams
+            ? unwrappedRequest.params
+            : new HttpParams({fromObject: unwrappedRequest.params}),
+        reportProgress: unwrappedRequest.reportProgress,
+        withCredentials: unwrappedRequest.withCredentials,
+        responseType,
+      },
```
optional: I could see refactoring some of this block into one or two smaller functions, e.g.

```ts
return new HttpRequest(
  unwrappedRequest.method ?? 'GET',
  unwrappedRequest.url,
  unwrappedRequest.body ?? null,
  normalizedRequestOptions(...),
);
```

jelbourn commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,460 @@
+        let resolve: ((value: Signal<ResourceStreamItem<T>>) => void) | undefined;
+        const promise = new Promise<Signal<ResourceStreamItem<T>>>((r) => (resolve = r));
+
+        function send(value: ResourceStreamItem<T>): void {
```
Why `function` instead of an arrow function? 
(add a comment?)

alxhub replied:
Why use an arrow function if closing over `this` isn't needed? :)

jelbourn replied:
I've just internalized the practice of using arrow functions inline as generally less error prone (no surprising `this`, no surprise hoisting)

jelbourn commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,460 @@
+          },
+        });
+
+        abortSignal.addEventListener('abort', () => sub.unsubscribe());
```
Is this `addEventListener` duplicated with the one above, or am I misreading this?

jelbourn commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,460 @@
+ *
+ * @experimental
+ */
+export interface HttpResourceRequest {
```
optional: I could see refactoring out the request-related types into a sibling file (`request.ts`)

alxhub commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,460 @@
+          resolve = undefined;
+        }
+
+        sub = this.client.request(request!).subscribe({
```
It's #58871 - should be fixed as soon as I have time to TGP the fix.

alxhub commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,460 @@
+  /**
+   * Transform the result of the HTTP request before it's delivered to the resource.
+   *
+   * `map` receives the value from the HTTP layer as its raw type (e.g. as `unknown` for JSON data).
+   * It can be used to validate or transform the type of the resource. For example, a Zod parser can
+   * be provided for `map`, which will result in the resource having a strongly typed value instead
+   * of `unknown`.
```
We want users to be able to narrow the type of the actual `Resource` instance, so that wherever you plug the resource (route resolvers, etc) you can pass your properly typed object.

Using a `computed` downstream would work for the value alone, but it'd be too late to do any type checking or inference on the whole resource object. Basically, `computed` returns a `Signal`, not a `Resource`. It's a separate question whether we want to have something like a `computedResource` / `linkedResource` for doing derivations while preserving the `Resource` type...

alxhub commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,460 @@
+export interface HttpResourceRequest {
+  url: string;
+  method?: string;
+  body?: unknown;
```
There's no point to a generic, since nothing depends on the type of the `body` in any way that could be typechecked or observed by the user.

alxhub commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,460 @@
+ */
+export const httpResource: HttpResourceFn = (() => {
+  const jsonFn = httpResourceImpl.bind(undefined, 'json') as unknown as HttpResourceFn;
+  jsonFn.arrayBuffer = httpResourceImpl.bind(undefined, 'arraybuffer') as any;
```
Excellent suggestion :)

alxhub commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,460 @@
+ */
+export interface HttpResourceFn {
+  /**
+   * Create a `Resource` that fetches data with an HTTP GET request to the given URL, which may be
+   * reactive.
```
Oh, the intention is to say that the URL input might be reactive, not that the resulting resource is.

cexbrayat commented on packages/common/http/src/resource.ts:
```diff
@@ -0,0 +1,417 @@
+  <TResult = unknown>(
+    url: string | (() => string | undefined),
+    options: HttpResourceOptions<TResult, unknown> & {defaultValue: NoInfer<TResult>},
+  ): HttpResource<TResult>;
```
ooc: is there a reason to name the returned value `HttpResource` instead of `HttpResourceRef` (as `ResourceRef` is returned by `resource()`)?

alxhub replied:
Actually I just renamed it to `HttpResourceRef` (before I saw this comment 😆)

cexbrayat commented on packages/common/http/src/resource_api.ts:
```diff
@@ -0,0 +1,135 @@
+  /**
+   * Signal of the response status code, when available.
+   */
+  readonly statusCode: Signal<number | undefined>;
```
is it on purpose to use number instead of`HttpStatusCode`?

alxhub replied:
Yes - `HttpStatusCode` is useful for comparisons, but we can't guarantee that the status code returned by the server will be present in the enum.

JeanMeche replied:
Also that enum should burn into flames. (It's the largest one we have)

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-core

thePunderWoman commented:
This PR was merged into the repository by commit 3e39da593a0a0c047a2a03b8d5fcabf9dbace40f.

The changes were merged into the following branches: main
